src/library/dtls.cpp

```diff
@@ -360,47 +404,48 @@ void DtlsSession::HandleEvent(short aFlags)
                 error = Read();
             }
         }
-        if ((aFlags & EV_WRITE) && error != Error::kTransportFailed)
+        if ((aFlags & EV_WRITE) && !ShouldStop(error))
         {
             error = TryWrite();
         }
         break;
 
-    case State::kDisconnecting:
-        Disconnect();
-        break;
-
     default:
         ASSERT(false);
         break;
     }
 
 exit:
-    if (error == Error::kTransportFailed)
+    if (ShouldStop(error))
     {
-        Disconnect();
+        Disconnect(error);
     }
 }
 
 Error DtlsSession::SetClientTransportId()
 {
     ASSERT(mIsServer && !mIsClientIdSet);
-    Error error = Error::kNone;
-    int   rval;
+    Error error    = Error::kNone;
+    int   rval     = 0;
     auto  peerAddr = GetPeerAddr();
 
     VerifyOrExit(peerAddr.IsValid(), error = Error::kInvalidAddr);
 
     rval = mbedtls_ssl_set_client_transport_id(&mSsl, reinterpret_cast<const uint8_t *>(&peerAddr.GetRaw()[0]),
                                                peerAddr.GetRaw().size());
-    VerifyOrExit(rval == 0, error = Error::kOutOfMemory);
+    SuccessOrExit(error = ErrorFromMbedtlsError(rval));
 
     mIsClientIdSet = true;
 
 exit:
     return error;
 }
 
+bool DtlsSession::ShouldStop(Error aError)
```

Comment:
Do you think its better to define a member method like `bool Error::IsTemporary()` or `bool Error::IsFatal()` and call it here?

Reply:
The error code is generic, whether an error code is fatal or not may differ in different context. So I think it is not applicable at here. thoughts?